[FEAT] 멘토링 승인/거절 시 SMS 발송 기능 구현

dye0p opened the pull request:
## Issue Number
closed #262

## As-Is
<!-- 문제 상황 정의 -->
- 현재는 멘토가 예약을 승인하더라도, 해당 예약의 멘티에게 아무런 알림이 가지 않습니다.
 - 멘티는 예약이 승인되었는지, 그리고 멘토의 연락처가 무엇인지 확인하기 위해 직접 애플리케이션에 접속하여 확인해야 하는 불편함이 있습니다.

## To-Be
<!-- 변경 사항 -->
- 멘토가 예약 상태를 승인 또는 거절로 변경하면, 멘티에게 결과(승인 시 멘토 연락처 포함)가 담긴 SMS 알림이 자동으로 전송되도록 기능을 개선합니다.

**1. Facade 서비스 도입 (`MentoringReservationFacadeService`)**
- 기존 `MentoringReservationService`의 이름을 `MentoringReservationFacadeService`로 변경하고, 역할을 재정의했습니다.
- 이 Facade 서비스는 예약 상태를 변경하는 ReservationService와 SMS를 발송하는 `SmsRestClientService`를 함께 호출하는 역할을 담당합니다. 이를 통해 상태 변경과 알림 발송이라는 두 가지 작업을 하나의 트랜잭션처럼 처리하면서, 각 서비스의 책임은 명확하게 분리했습니다.

**2. 알림 로직 구현**
- `ReservationController`는 이제 상태 변경 요청 시 `MentoringReservationFacadeService`의 `updateStatusAndSendSms` 메서드를 호출합니다.
- `Status` Enum에 `isApprove()`, `isReject()`, `isNotifiable()`과 같은 헬퍼 메서드를 추가하여, 상태 값에 따른 비즈니스 로직을 더 명확하고 안전하게 처리하도록 개선했습니다.
- `SmsMessageFormatter`에 예약 승인/거절 시나리오에 맞는 별도의 메시지 포맷을 추가했습니다.

**3. 안정성 및 예외 처리 강화**
- 문자열로 받은 상태 값을 안전하게 `Status` Enum으로 변환하기 위해 `Status.of()` 팩토리 메서드를 추가했습니다.
- 유효하지 않은 상태 값이 들어올 경우 `NotFoundStatusException` 예외를 발생시키고, 이를 처리하는 핸들러를 `GlobalExceptionHandler`에 추가하여 시스템 안정성을 높였습니다.

**4. 테스트 코드 수정**
- ReservationControllerTest 통합 테스트에서 @MockBean을 사용해 SmsRestClientService를 Mocking 처리했습니다. 
- 이를 통해 실제 SMS가 발송되는 것을 막고, 외부 API 의존성 없이 안정적으로 테스트를 수행할 수 있도록 변경했습니다.
- Status Enum에 추가된 of() 메서드에 대한 단위 테스트를 추가하여 신뢰도를 높였습니다.


## Check List
- [x] 테스트가 전부 통과되었나요?
- [x] 모든 commit이 push 되었나요?
- [x] merge할 branch를 확인했나요?
- [x] Assignee를 지정했나요?
- [x] Label을 지정했나요?
- [x] 닫을 이슈 번호를 지정했나요?


## (Optional) Additional Description

### 우려 사항
현재 `MentoringReservationFacadeService`는 단순히 다른 서비스들을 호출하는 것을 넘어, 어떤 조건에서 알림을 보낼지(`status.isNotifiable()`) 그리고 어떤 내용의 메시지를 만들지(`createMessage()`) 결정하는 비즈니스 로직을 일부 포함하고 있습니다.

이는 Facade의 역할을 다소 넘어서는 것으로, 향후 이메일 알림 등 새로운 기능이 추가될 경우 Facade가 점점 더 많은 책임을 갖게 되어 유지보수가 어려워질 수 있습니다.

개선 방안으로는 향후 리팩토링 시 `Spring Application Event`를 도입하는 것을 고려해볼 수 있습니다.

dye0p committed:
feat: 예약 승인/거절에 보낼 메세지 포맷 추가

dye0p committed:
feat: 예약의 상태가 승인/거절 상태로 변경되면 sms을 발송하는 기능 구현

dye0p committed:
refactor: 예약 상태 변경 기능과 sms 전송 기능 트랜잭션 분리

- 예약 변경과 sms 전송 로직이 하나의 트랜잭션에 묶여있어, sms 전송이 실패하면 상태 변경까지 롤백되는 상황이 문제였음. 두가지 로직을 파사드 객체에 위임하여 트랜잭션 분리

dye0p committed:
refactor: MentoringReservationService 객체 네이밍 변경(MentoringReservationService -> MentoringReservationFacadeService)

- MentoringReservationService 객체는 Facade 역할을 하는 객체이므로 네이밍을 명확하게 변경

dye0p committed:
refactor: 매개변수 네이밍 명확하게 변경

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Status.java:
```diff
@@ -19,6 +20,16 @@ public enum Status {
         this.value = value;
     }
 
+    public static Status of(String updateStatus) {
```
p3: 매개변수 이름을 `value`로 받는 건 어떤가요? `value`로 enum을 찾는 것 같아요!

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringReservationFacadeService.java:
```diff
@@ -30,4 +31,34 @@ public ReservationCreateResponse reserveMentoring(ReservationCreateDto dto) {
+        sendSms(reservation, status);
+    }
+
+    private void sendSms(Reservation reservation, String updateStatus) {
```
p3: `sendStatusUpdateSms()`와 같이 상태가 변경되었다는 문자를 보낸다는 의미를 담는 메서드 네이밍은 어떤가요?? 뭔가 지금은 단순히 sms를 보낸다는 의미인 것 같습니다!

ppparkta replied:
p3: 변수명 `updatedStatus` 어떠신가요?

dye0p replied:
반영하겠습니다!

[refactor: status 매개변수 네이밍 명확하게 변경(status -> updatedStatus)](https://github.com/woowacourse-teams/2025-Fit-toring/pull/331/commits/c6a16945e27b3743d21bf208442846781e493c20)

kimdevv commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringReservationFacadeService.java:
```diff
@@ -30,4 +31,34 @@ public ReservationCreateResponse reserveMentoring(ReservationCreateDto dto) {
+        String context = reservation.getContext();
+        String mentorPhoneNumber = reservation.getMentorPhone();
+
+        if (status.isNotifiable()) {
```
p3: 이 `notifiable`은 어떤 걸 의미하나요????

dye0p replied:
"**고지 가능**" 이라는 의미를 담으려고 했습니다! 의미가 모호하게 느껴지셨군요..!

kimdevv reviewed (approved):
고생하셨습니다~~~ 궁금한 점 커멘트로 남겼습니다!

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Status.java:
```diff
@@ -28,6 +39,18 @@ public void validate(Status updateStatus) {
+        return this.isApprove() || this.isReject();
+    }
+
+    public boolean isApprove() {
+        return this.equals(APPROVE);
+    }
+
+    public boolean isReject() {
+        return this.equals(REJECT);
+    }
```
p2 : 의미 상 isApproved, isRejected 는 어떨까요?

dye0p replied:
상수명이 변경되었으니 변경되는것이 좋아보입니다!

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/business/model/Status.java:
```diff
@@ -28,6 +39,18 @@ public void validate(Status updateStatus) {
         }
     }
 
+    public boolean isNotifiable() {
+        return this.isApprove() || this.isReject();
+    }
```
p2 : isNotifiable 보다는 shouldNotify 는 어떨까요?

dye0p replied:
가능/불가능의 의미인데, 적절할까요? 영어에 약해서 죄송합니다..

jye326 replied:
저는 '고지할 수 있는' 보다 '고지해야 하는' 이 역할에 더 맞는 네이밍이라고 생각합니다!

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringReservationFacadeService.java:
```diff
@@ -30,4 +31,34 @@ public ReservationCreateResponse reserveMentoring(ReservationCreateDto dto) {
         return ReservationCreateResponse.from(reservation);
     }
+
+    public void updateStatusAndSendSms(Long reservationId, String status) {
```
p2 : 예약의 상태를 변경함을 드러내기 위해 updateReservationStatusAndSendSms 는 어떨까요?

dye0p replied:
변경하는게 더 명확하겠네요! 반영하겠습니다!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/infra/SmsMessageFormatter.java:
```diff
@@ -9,6 +9,38 @@ public class SmsMessageFormatter {
+
+    public String createRejectReservationMessage(String mentorName, String context) {
+        return String.format("""
+                        [핏토링] 멘토링 예약이 거절되었습니다.
+                        
+                        멘토: %s
+                        상담 내용: %s
+                        
+                        아쉽게도 이번 예약은 진행되지 않게 되었어요.
+                        다른 멘토링을 신청해 보시는 건 어떨까요?
```
p3: 기획쪽에서 의견 있는데 예약 취소 시 subject와 상담 내용을 보내지 않는 거 어떠신가요?

> [핏토링] %s님과의 멘토링이 취소되었습니다. 
> 마이페이지에서 자세한 내용을 확인하실 수 있습니다.

이런식으로 자세한 내용을 확인하기 위해서는 마이페이지로 유도하는 식으로 수정되면 좋을거 같아요. 장문 sms 전송 비용부담 + 일부 문장이 자칫하면 사용자에게 불쾌함을 느끼게 할 수 있을 것 같아요!

dye0p replied:
사용자 배려 좋네요! 반영하겠습니다!

[fix: 예약 거절시 문자 내용 변경](https://github.com/woowacourse-teams/2025-Fit-toring/pull/331/commits/27cc9692d603478ace5a04b29b1bd08fb015ac5a)

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/infra/SmsMessageFormatter.java:
```diff
@@ -9,6 +9,38 @@ public class SmsMessageFormatter {
     private static final String VERIFICATION_MESSAGE_PREFIX = "핏토링 인증번호는 [";
     private static final String VERIFICATION_MESSAGE_SUFFIX = "] 입니다.";
 
+    public String createApproveReservationMessage(String mentorName, String context, String mentorPhone) {
```
p3 : `createApproveReservationMessage `-> `approvedReservationMessage `은 어떨까요?
- 클래스가 포매터기 때문에 create 접두사 없어도 이해할 수 있을것 같습니다.
- Approve 보다 Approved 가 좋을 것 같습니다.

dye0p replied:
좋습니다! 반영하겠습니다.

[refactor: SmsMessageFormatter 메서드 네이밍 변경](https://github.com/woowacourse-teams/2025-Fit-toring/pull/331/commits/ee5597633d2b8869f501d3c7a03c1fd5f49fff8c)

jye326 commented on backend/fittoring/src/main/java/fittoring/mentoring/infra/SmsMessageFormatter.java:
```diff
@@ -9,6 +9,38 @@ public class SmsMessageFormatter {
+        );
+    }
+
+    public String createRejectReservationMessage(String mentorName, String context) {
```
p3 : 같은 이유로 `rejectedReservationMessage()` 가 좋을 것 같습니다!

dye0p replied:
좋습니다!

[refactor: SmsMessageFormatter 메서드 네이밍 변경](https://github.com/woowacourse-teams/2025-Fit-toring/pull/331/commits/ee5597633d2b8869f501d3c7a03c1fd5f49fff8c)

jye326 reviewed (commented):
굳~ 코멘트 남겼습니다! 확인해주세요~!

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringReservationFacadeService.java:
```diff
@@ -30,4 +31,34 @@ public ReservationCreateResponse reserveMentoring(ReservationCreateDto dto) {
+        }
+        return smsMessageFormatter.createRejectReservationMessage(mentorName, context);
+    }
+
```
p2: ㄱㅎ

dye0p replied:
네?..

dye0p replied:
"개행" 맞나요?

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringReservationFacadeService.java:
```diff
@@ -30,4 +31,34 @@ public ReservationCreateResponse reserveMentoring(ReservationCreateDto dto) {
+        sendSms(reservation, status);
+    }
+
+    private void sendSms(Reservation reservation, String updateStatus) {
```
p3: `sendSms()`의 위치가 적절하지 않아보여요. 

> 현재 MentoringReservationFacadeService는 단순히 다른 서비스들을 호출하는 것을 넘어, 어떤 조건에서 알림을 보낼지(status.isNotifiable()) 그리고 어떤 내용의 메시지를 만들지(createMessage()) 결정하는 비즈니스 로직을 일부 포함하고 있습니다.
> 이는 Facade의 역할을 다소 넘어서는 것으로, 향후 이메일 알림 등 새로운 기능이 추가될 경우 Facade가 점점 더 많은 책임을 갖게 되어 유지보수가 어려워질 수 있습니다.

reservation의 상태가 메시지를 결정한다면 메시지를 생성하는 것은 `ReservationService`의 책임 아닐까요? 해당 서비스가 무거워진다면 고려해주신 알림을 차용해서 `ReservationNotificationService` 정도로 분리할 수 있을 것 같아요. 그러면 본문에 남겨주신 책임 분리 문제도 어느정도 해결할 수 있을 것 같습니다.

별개로 이벤트 도입되면 좋겠네요!

dye0p replied:
> ReservationService의 책임 아닐까요?  해당 서비스가 무거워진다면 고려해주신 알림을 차용해서 ReservationNotificationService 정도로 분리할 수 있을 것 같아요.

처음엔 ReservationServic에 해당 로직을 넣으려고 했는데, 조금 생각해보니 "메세지를 전송한다"라는 책임은 예약에서만의 책임이라고 보기엔 다른 도메인에서도 사용될 수 있다고 생각했습니다. 지금은 sendSms()라는 로직의 내부 로직을 그대로 사용하는 것 같은 느낌인데, 말씀 해주신 것 처럼 `ReservationNotificationService` 같이 예약과 관련된 sms 전송 책임을 가지는 객체로 분리하는것이 나을 것 같아요! 

> 별개로 이벤트 도입되면 좋겠네요!

동의합니다! 추후에 sms 전송을 이벤트 기반으로 고도화 하면 좋을 것 같아요!

[refactor: 예약 관련 sms 전송로직 별도의 service 객체로 분리 + 메서드 네이밍 명확하게 수정](https://github.com/woowacourse-teams/2025-Fit-toring/pull/331/commits/bc7797fe09d8cabcafe69a1798c1138661e14574)

ppparkta commented on backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringReservationFacadeService.java:
```diff
@@ -30,4 +31,34 @@ public ReservationCreateResponse reserveMentoring(ReservationCreateDto dto) {
+
+        if (status.isNotifiable()) {
+            String message = createMessage(status, mentorName, context, mentorPhoneNumber);
+            Phone menteePhone = reservation.getMentee().getPhone();
```
p3: reservation에서 멘티의 폰넘버를 반환하는 메서드를 만드는거 어떠신가요?

dye0p replied:
[[[FEAT] 멘토용 예약 관리(예약 조회/예약 상태 변경/예약자 전화번호 조회) API 구현](https://github.com/woowacourse-teams/2025-Fit-toring/pull/316#top)](https://github.com/woowacourse-teams/2025-Fit-toring/pull/316)

해당 PR에서 반영하였습니다!

ppparkta reviewed (commented):
고생하셨어요! 글로 잘 정리해주셔서 리뷰하기 너무 편했네요. 감사합니다 😄

dye0p committed:
refactor: 예약의 상태를 변경하고 sms를 전송하는 메서드 네이밍 명확하게 변경

dye0p committed:
refactor: SmsMessageFormatter 메서드 네이밍 변경

- Formatter 라는 클래스 네이밍으로 '생성'의 의미를 담고 있기에 메서드에 create 라는 접두사 제거

dye0p committed:
fix: 예약 거절시 문자 내용 변경

- 자세한 내용을 확인하기 위해서 사용자를 마이페이지로 유도하기 위해서 불필요한 내용 제거
- 장문의 SMS 전송시 비용부담의 문제로 필요한 문장만 포함시킴

dye0p committed:
refactor: 예약 완료, 인증번호 전송 메세지 생성 메서드 네이밍 변경

dye0p committed:
refactor: 예약 관련 sms 전송로직 별도의 service 객체로 분리 + 메서드 네이밍 명확하게 수정

- 기존 Facade 객체에서 sms 전송 관련 로직에 직접 관여하는 것이 책임이 아니라 판단하여 별도의 로직을 수행하는 service객체로 분리하여 Facade 객체에서는 service 로직의 호출만 담당할 수 있도록 변경

dye0p committed:
refactor: status 매개변수 네이밍 명확하게 변경(status -> updatedStatus)

- 변경된 상태라는 것을 명확하게 하기 위해서 변경

dye0p reviewed (commented):
피드백을 반영하고 수정하였습니다! 검토 부탁드립니다~

dye0p committed:
Merge branch 'feature/260-mentor-reservations-manager' into feature/reservation-approve-send-sms

# Conflicts:
#	backend/fittoring/src/main/java/fittoring/exception/GlobalExceptionHandler.java
#	backend/fittoring/src/main/java/fittoring/mentoring/business/service/MentoringReservationService.java
#	backend/fittoring/src/main/java/fittoring/mentoring/infra/SmsMessageFormatter.java

ppparkta commented:
고생하셨습니다. 전체적으로 개행 관련 컨벤션 적용 후 재요청 주시면 바로 승인하겠습니다.
주말에도 고생하는 멋진 팀원이 있어서 든든하네요 👍

dye0p committed:
refactor: 개행 리팩터링

dye0p reviewed (commented):
개행 부분 리팩터링 했는데, 구체적으로 어느 부분을 말씀하신건지 이해를 못해서 우선 임의로 했습니다!

jye326 reviewed (approved):
주말인데도 열심히 작업해주셨네요! 👍👍 승인하겠습니다~
(슬쩍 코멘트 한 번만 확인해주세요~)

ppparkta reviewed (approved):
개행 맞습니다. 승인하겠습니다. 고생많으셨어요~~

dye0p committed:
Merge branch 'develop' into feature/reservation-approve-send-sms

# Conflicts:
#	backend/fittoring/src/main/java/fittoring/mentoring/business/model/Status.java
